src/pstore/pstore.c

```diff
@@ -0,0 +1,517 @@
+    char *contents;
+    LIST_FIELDS(struct pstore_entry_s, dll); /* dll doubly linked list */
+} pstore_entry_t;
+
+static void sort_files(pstore_entry_t **list, int ascending) {
+        /* Simple brute force linear sort */
+        int sorted = 0;
+        pstore_entry_t *pe, *tail;
+        do {
+                sorted = 1;
+
+                LIST_FIND_TAIL(dll, *list, tail);
+                LIST_FOREACH(dll, pe, *list) {
+                        if (ascending) {
+                                if (!(pe == tail) && /* not end of list */
+                                        /* lexigraphical sort, ascending */
+                                        (strcmp(pe->direntbuf.d_name, pe->dll_next->direntbuf.d_name) > 0)) {
+                                        LIST_REMOVE(dll, (*list), pe);
+                                        LIST_APPEND(dll, (*list), pe);
+                                        sorted = 0;
+                                        break; /* re-start sort */
+                                }
+                        }
+                        else {
+                                if (!(pe == *list) && /* not start of list */
+                                        /* lexigraphical sort, descending */
+                                        (strcmp(pe->direntbuf.d_name, pe->dll_prev->direntbuf.d_name) > 0)) {
+                                        LIST_REMOVE(dll, (*list), pe);
+                                        LIST_PREPEND(dll, (*list), pe);
+                                        sorted = 0;
+                                        break; /* re-start sort */
+                                }
+                        }
+                }
+        } while (!sorted);
+}
+
+static void free_files(pstore_entry_t **list) {
+        pstore_entry_t *pe;
+        LIST_FOREACH(dll, pe, *list) {
+                free(pe->contents);
+                LIST_REMOVE(dll, (*list), pe);
+                free(pe);
+        }
+}
+
+static int move_file(pstore_entry_t *pe, const char *subdir) {
+        _cleanup_free_ char *ifd_path;
+        int r = 0;
+
+        ifd_path = path_join(arg_sourcedir, pe->direntbuf.d_name);
+        if (!ifd_path)
+                return log_oom();
+
+        if (arg_storage == PSTORE_STORAGE_EXTERNAL) {
+                /* This code copies the file from the pstore to other storage.
+                 * The rename() syscall is not utilized as it results in the
+                 * Invalid cross-device link error.
+                 * In addition, an optional subdirectory can be specified in
+                 * forming the final destination path.  */
+                _cleanup_free_ char *ofd_path;
+                ofd_path = path_join(arg_archivedir, subdir, pe->direntbuf.d_name);
+                if (!ofd_path)
+                        return log_oom();
+
+                /* Make sure destination exists */
+                r = mkdir_parents(ofd_path, 0755);
+                if (r == 0) {
```

Comment:
Return earlier, and use negative errno `r` instead of `errno`:
```
if (r < 0)
        return log_error_errno(r, "...");
```

Reply:
done